File: tests/e2e/scripts/array_methods.ts
function main(): void {
    const nums: number[] = [1, 2, 3, 4, 5];

    // map: verify by reducing the result
    const doubled = nums.map((x: number): number => x * 2);
    console.log("map length:", doubled.length);
    const doubledSum = doubled.reduce((acc: number, x: number): number => acc + x, 0);
    console.log("map sum:", doubledSum);

    // filter: verify length and sum
    const evens = nums.filter((x: number): boolean => x % 2 === 0);
    console.log("filter length:", evens.length);
    const evensSum = evens.reduce((acc: number, x: number): number => acc + x, 0);
    console.log("filter sum:", evensSum);

    // some
    console.log("some >3:", nums.some((x: number): boolean => x > 3));
    console.log("some >10:", nums.some((x: number): boolean => x > 10));

    // every
    console.log("every >0:", nums.every((x: number): boolean => x > 0));
    console.log("every >3:", nums.every((x: number): boolean => x > 3));

    // reduce
    const sum = nums.reduce((acc: number, x: number): number => acc + x, 0);
    console.log("reduce sum:", sum);

    // forEach
    let total: number = 0;
    nums.forEach((x: number): void => {
        total = total + x;
    });
    console.log("forEach total:", total);

    // indexOf
    console.log("indexOf 3:", nums.indexOf(3));
    console.log("indexOf 99:", nums.indexOf(99));

    // slice: verify length and sum
    const sliced = nums.slice(1, 3);
    console.log("slice length:", sliced.length);
    const slicedSum = sliced.reduce((acc: number, x: number): number => acc + x, 0);
    console.log("slice sum:", slicedSum);

    // sort: verify by reducing sorted array
    const unsorted: number[] = [3, 1, 4, 1, 5];
    unsorted.sort();
    const sortCheck = unsorted.some((x: number): boolean => x === 1);
    console.log("sort contains 1:", sortCheck);
    console.log("sort length:", unsorted.length);

    // find: returns T | undefined in TS, Option<T> in Rust
    const firstOver3 = nums.find((x: number): boolean => x > 3);
    if (firstOver3 !== undefined) {
        console.log("find >3:", firstOver3);
    } else {
        console.log("find >3: none");
    }
    const firstOver100 = nums.find((x: number): boolean => x > 100);
    if (firstOver100 !== undefined) {
        console.log("find >100:", firstOver100);
    } else {
        console.log("find >100: none");
    }

    // filter with captured variable (threshold) — verifies filter closure
    // receives &T (Rust Iterator::filter semantics) but body uses T by value
    const threshold: number = 2;
    const above = nums.filter((x: number): boolean => x > threshold);
    console.log("filter >threshold length:", above.length);
    const aboveSum = above.reduce((acc: number, x: number): number => acc + x, 0);
    console.log("filter >threshold sum:", aboveSum);
}

File: tests/e2e/scripts/array_methods.rs
fn main() {
    let mut nums: Vec<f64> = vec![1.0, 2.0, 3.0, 4.0, 5.0];
    let doubled = nums
        .iter()
        .map(|x: f64| -> f64 { x * 2.0 })
        .collect::<Vec<_>>();
    println!("{} {}", "map:", doubled);
    let evens = nums
        .iter()
        .filter(|x: f64| -> bool { x % 2.0 == 0.0 })
        .collect::<Vec<_>>();
    println!("{} {}", "filter:", evens);
    let found = nums.iter().find(|x: f64| -> bool { x > 3.0 });
    println!("{} {}", "find:", found);
    println!(
        "{} {}",
        "some >3:",
        nums.iter().any(|x: f64| -> bool { x > 3.0 })
    );
    println!(
        "{} {}",
        "some >10:",
        nums.iter().any(|x: f64| -> bool { x > 10.0 })
    );
    println!(
        "{} {}",
        "every >0:",
        nums.iter().all(|x: f64| -> bool { x > 0.0 })
    );
    println!(
        "{} {}",
        "every >3:",
        nums.iter().all(|x: f64| -> bool { x > 3.0 })
    );
    let sum = nums.iter().fold(0.0, |acc, x| -> f64 { acc + x });
    println!("{} {}", "reduce sum:", sum);
    let mut total: f64 = 0.0;
    nums.iter().for_each(|x: f64| -> () {
        total = total + x;
    });
    println!("{} {}", "forEach total:", total);
    println!(
        "{} {}",
        "indexOf 3:",
        nums.iter()
            .position(|item| *item == 3.0)
            .map(|i| i as f64)
            .unwrap_or(-1.0)
    );
    println!(
        "{} {}",
        "indexOf 99:",
        nums.iter()
            .position(|item| *item == 99.0)
            .map(|i| i as f64)
            .unwrap_or(-1.0)
    );
    let sliced = nums[1..3].to_vec();
    println!("{} {}", "slice 1,3:", sliced);
    let mut unsorted: Vec<f64> = vec![3.0, 1.0, 4.0, 1.0, 5.0];
    unsorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    println!("{} {}", "sort:", unsorted);
}
